pkg/scheduler/plugins/random/calculate.go

```diff
@@ -0,0 +1,115 @@
+package random
+
+import (
+	v1 "k8s.io/api/core/v1"
+	"math"
+	"volcano.sh/volcano/pkg/scheduler/api"
+)
+
+func (rp *randomPlugin) Calculate() {
+	guarantees := make([]*api.Resource, 0)
+	capacities := make([]*api.Resource, 0)
+	requests := make([]*api.Resource, 0)
+	weights := make([]int32, 0)
+	queueMap := make(map[int]api.QueueID)
+	queueIndex := 0
+	var totalWeight int32
+	for queueId, attr := range rp.queueOpts {
+		guarantees = append(guarantees, attr.guarantee)
+		capacities = append(capacities, attr.realCapability)
+		requests = append(requests, attr.request)
+		weights = append(weights, attr.weight)
+		totalWeight += attr.weight
+		queueMap[queueIndex] = queueId
+		queueIndex++
+	}
+	totalResource := rp.totalResource.Clone()
+	weightResources := make([]*api.Resource, 0)
+	for _, w := range weights {
+		weightResources = append(weightResources, totalResource.Multi(float64(w)/float64(totalWeight)))
+	}
+
+	deserves := calculate(rp.totalResource, rp.totalGuarantee, guarantees, capacities, requests, weightResources,
+		len(rp.queueOpts), rp.tryNum, rp.weightCostWeight, rp.requestCostWeight)
+	for i, d := range deserves {
+		queueId := queueMap[i]
+		rp.queueOpts[queueId].deserved = d
+	}
+}
+
+func calculate(total, totalGuarantees *api.Resource, guarantees, capacities, requests, weightResources []*api.Resource,
+	queueNum, tryNum, weightCostWeight, requestCostWeight int) []*api.Resource {
+
+	remaining := total.Clone().Sub(totalGuarantees)
+	min := math.MaxFloat64
+	var d []*api.Resource
+	// find the smallest cost in tryNum tries
+	for i := 0; i < tryNum; i++ {
+		// try with random split
+		deserves := try(remaining.Clone(), guarantees, capacities, queueNum)
+		// calculate cost
+		costw := costs(deserves, weightResources)
```

Comment:
As what we discussed in the weekly meeting, it is better to expose an interface or callback function for users to define custom `costs` function. It will be more flexible to adjust different scenarios.

Reply:
done